### PR TITLE
add a random delay to the various pollers

### PR DESCRIPTION
**Pull request by alucryd**
## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragments` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

I have a buildbot with hundreds of git pollers, running all of them at the same time hogs my CPU and disks with iowait. This is intended to spread the load over a period of time, and does wonders on my poor server now.

### CODE REVIEW

**Comment by alucryd** [comment]
All should be good now, the last error seems unrelated: https://buildbot.buildbot.net/#/builders/54/builds/964

File: master/buildbot/util/poll.py [hunk 1]
```diff
@@ -37,8 +39,10 @@ def __init__(self, fn, instance, reactor):
         self._reactor = reactor
 
     @defer.inlineCallbacks
-    def _run(self):
+    def _run(self, random_delay=0):
         self.running = True
+        if random_delay:
+            yield task.deferLater(self._reactor, randint(0, random_delay), lambda: None)
```
Comment (rajdeepbharati) [inline master/buildbot/util/poll.py:45]: Wouldn't it also be nice to have an option to set minimum delay>0 ?

Comment (alucryd) [reply master/buildbot/util/poll.py:45]: Absolutely, I'll add this as well

File: master/docs/manual/configuration/changesources.rst [hunk 2]
```diff
@@ -599,6 +599,20 @@ It accepts the following arguments:
     How often to poll, in seconds.
     Defaults to 600 (10 minutes).
 
+``pollRandomDelayMin``
+    minimum delay in seconds to wait before each poll, default is 0
```
Comment (p12tic) [inline master/docs/manual/configuration/changesources.rst:603]: Missing dot.

File: master/docs/manual/configuration/changesources.rst [hunk 3]
```diff
@@ -599,6 +599,20 @@ It accepts the following arguments:
...
+    This is useful in case you have a lot of pollers and you want to spread the
+    polling load over a period of time.
+    Setting it equal to the maximum delay will effectively delay all polls by a
+    fixed amount of time.
+    Must be less than or equal to the maximum delay.
+
+``pollRandomDelayMax``
+    maximum delay in seconds to wait before each poll, default is 0
```
Comment (p12tic) [inline master/docs/manual/configuration/changesources.rst:611]: Missing dot.

File: master/buildbot/changes/base.py [hunk 4]
```diff
@@ -55,24 +55,59 @@ def _unclaimService(self):
 class ReconfigurablePollingChangeSource(ChangeSource):
     pollInterval = None
     pollAtLaunch = None
-
-    def checkConfig(self, name=None, pollInterval=60 * 10, pollAtLaunch=False):
+    pollRandomDelayMin = None
+    pollRandomDelayMax = None
+
+    def checkConfig(
```
Comment (p12tic) [inline master/buildbot/changes/base.py:61]: Can we keep aligning the arguments with the opening parentheses like in the rest of the codebase?

Comment (alucryd) [reply master/buildbot/changes/base.py]: Sorry about that, I come from black so I didn't think twice about this. Moved back to inline while staying under the 100 mark.

**Comment by alucryd** [comment]
> Formatting needs to be brought back inline with the rest of the codebase. Otherwise looks good to me, thanks!

Sorry about that, I come from black so I didn't think twice about this. Moved back to inline while staying under the 100 mark.

**Pull request merged**